### ISSUE TITLE
Howto process further with Match.orElse()

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#227]
On the one hand it is common to use `otherwise`. On the other hand, `orElse()` is unified throughout Java and Javaslang.

We are not able to let `Match` fully implement `Valences.Univalent`. However, `Match` should implement:

```java
T orElse(T other);
T orElseGet(Supplier<? extends T> other);
<X extends Throwable> T orElseThrow(Supplier<X> exceptionSupplier) throws X;
```

**Comment by danieldietrich** [comment]
After thinking again, I also tend to use `otherwise()`. There is nothing to `get()` from a `Match` before applying it to an `Object`.

**Comment by jbgi** [comment]
You could also design match to use
```java
Number i = Integer.valueOf(1);

BigDecimal asBigDecimal =
Match.of(i)
     .when((Integer i) -> new BigDecimal(i))
     .when((BigDecimal bd) -> bd)
     .getOrlElse(BigDecimal.ZERO);

Option<BigDecimal> asBigDecimalOption =
Match.of(i)
     .when((Integer i) -> new BigDecimal(i))
     .when((BigDecimal bd) -> bd)
     .get();
```
also only Function<? extends Number, R> should be allowed as when argument, IMO.
and the first when argument determine the type of R. (that would require a smarter Match builder)

**Comment by danieldietrich** [comment]
Wow, this would be exactly the solution in order to make the `when`-part of `Match` implement [Univalent](https://github.com/javaslang/javaslang/blob/master/src/main/java/javaslang/control/Valences.java) (difference: we call `toOption()` instead of `get()` then). Awesome!

The existing `Match` syntax allows us to match arbitrary objects (like Scala). Additionally it makes `Match` reusable (see example below) and also composable, because it is a `Function`.

Do you think that `Match` will be too complicated, if there are two ways to use it: (1) as value, (2) as function ? Simplicity is very important. Ideally there is only one way to achieve a goal. In case (1) I would use `orElse()`, in case (2) I would use `otherwise()` for the default case.

```java
class Test {

    static final Match<BigDecimal> TO_BIG_DECIMAL = Match
        .when((Integer i) -> new BigDecimal(i))
        .when((BigDecimal bd) -> bd)
        .otherwise(BigDecimal.ZERO);

    List<BigDecimal> mapToBigDecimal(List<Number> numbers) {
        return numbers.map(TO_BIG_DECIMAL);
    }
}
```

**Comment by danieldietrich** [comment]
Also I think of changing the method name `ofType()` to `as()` in order to override type inference for the return type of the `Match` if the return type of the first `when()` is not sufficient.

**Example:**

This does/will not compile:

```java
Match   .when((int i) -> 1)
        .when((double d) -> 1.0)
        .otherwise(-1)
        .apply(1);

Match.of(1)
        .when((int i) -> 1)
        .when((double d) -> 1.0)
        .orElse(-1);
```

This does/will compile:

```java
Match.as(Number.class)
        .when((int i) -> 1)
        .when((double d) -> 1.0)
        .otherwise(-1)
        .apply(1);

Match.of(1).as(Number.class)
        .when((int i) -> 1)
        .when((double d) -> 1.0)
        .orElse(-1);
```

**Comment by danieldietrich** [comment]
@jbgi I've prototyped your suggestion and the API feels really great! Here is the code (currently only within my fork): [Match2](https://github.com/danieldietrich/javaslang/blob/issue227/src/main/java/javaslang/control/Match2.java) \- And, oh my, this is most probably also a Monad :-)

Compared to the current [Match](https://github.com/javaslang/javaslang/blob/master/src-gen/main/java/javaslang/control/Match.java) solution there are the following changes:

* functions with primitive argument cannot be passed anymore to `when()`
* `Match` is no `Function` any more, i.e. it needs to be instantiated when applying a value: `list.map(x -> Match.of(x).when(...).get())`

The first point is not relevant in my opinion. The second point has an impact on performance, because the current reflection \*magic\* used to get the runtime type of the function argument is relatively slow. I use a trick here and deserialize lambdas to get the `MethodType`. When a match is executed a hundred thousand times a second, there will be a significant difference between the current and the new solution.

I'm a unsure if we should keep the current solution. I want to keep things simple and give only one option to use a `Match`...

**Comment by danieldietrich** [comment]
@jbgi We could do it as everyone else and just give a type hint to circumvent the reflection \*magic\*: `Match.of(i).when(Integer.class, i -> 1)`

**Comment by danieldietrich** [comment]
I played around with the new API. We see type erasure when using generic types:

```java
Match.of(List.of(1, 2, 3))
        .when(List.Nil.class, nil -> 0)
        .when(List.Cons.class, cons -> ((List<Integer>) cons).sum()) // cast is unsafe here!
        .get();
```

The previous approach was unsafe by design:

```java
Match   .when((Some<Integer> some) -> 1) // no cast needed, but also unsafe!
        .when((Some<String> some) -> Integer.parseInt(some.get()))
        .apply(new Some<>("123")); // result is 1 (!)
```

**Comment by danieldietrich** [comment]
Interesting: Intellij Idea compiles this, but Java 1.8.0_45 does not:

```java
Match.of(List.of(1, 2, 3)).as(Number.class)
        .when(List.Nil.class, nil -> 1.0)
        .when(List.Cons.class, cons -> (Integer) cons.head())
        .get();
```

Errors:

```
[ERROR] /Users/daniel/git/javaslang/src/test/java/javaslang/control/Match2Test.java:[35,62] cannot find symbol
  symbol:   method head()
  location: variable cons of type java.lang.Object
[ERROR] /Users/daniel/git/javaslang/src/test/java/javaslang/control/Match2Test.java:[36,21] incompatible types: java.lang.Object cannot be converted to java.lang.Number
```

**Comment by danieldietrich** [comment]
I've though about loosing generic types. The API `Match.of(value).when(Class, Function)` is not helpful. I will provide `Match.of(value).when(Function)` instead.

As mentioned already it is slower but keeping the current API `Match.when(Function)` can be used in performance critical cases. The `when(Function)` notation is then also concise for both types of Match builders (_value_ and _function_).

**Comment by danieldietrich** [comment]
Also required: Instead of

```java
final A result = Try.of(this::bunchOfWork)
    .recover(x -> Match
        .caze((Exception_1 e) -> ...)
        .caze((Exception_2 e) -> ...)
        .caze((Exception_n e) -> ...)
        .apply(x))
    .orElse(other);
```

we want to write

```java
final A result = Try.of(this::bunchOfWork)
    .recover(Match
        .caze((Exception_1 e) -> ...)
        .caze((Exception_2 e) -> ...)
        .caze((Exception_n e) -> ...))
    .orElse(other);
```

I.e. a `Match.Case` is a `Function<T, R>`, where `T` is defined implicitly by the context.

**Comment by martin-g** [comment]
@danieldietrich Maybe you have noticed it, maybe not - https://twitter.com/ScalaFact/status/606337216571260928. You can steal some ideas if there is something better ;-)

**Comment by danieldietrich** [comment]
@martin-g thanks, I will dive into it :+1:

**Comment by danieldietrich** [comment]
I've also experimented with 'emulating' pattern matching like [motif](https://github.com/johnlcox/motif) does. This is not practicable for Java developers. I already removed `ValueObject` from Javaslang, which should have been the base for `unapply`ing objects as basis for pattern matching. This is more a playground to see what is possible than being practicable in real world apps. It is better to not try this with Java. Javaslang's current pattern matching by type is useful and sufficient.

Thanks for the link! It was insightful.
